crates/stages/src/pipeline/mod.rs

```diff
@@ -83,57 +86,43 @@ impl<DB: Database, U: SyncStateUpdater> Default for Pipeline<DB, U> {
         Self {
             stages: Vec::new(),
             max_block: None,
-            events_sender: MaybeSender::new(None),
+            listeners: PipelineEventListeners::default(),
             sync_state_updater: None,
         }
     }
 }
 
 impl<DB: Database, U: SyncStateUpdater> Debug for Pipeline<DB, U> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Pipeline").field("max_block", &self.max_block).finish()
+        f.debug_struct("Pipeline")
+            .field(
+                "stages",
+                &self.stages.iter().map(|stage| stage.stage.id()).collect::<Vec<StageId>>(),
+            )
+            .field("max_block", &self.max_block)
+            .finish()
     }
 }
 
 impl<DB: Database, U: SyncStateUpdater> Pipeline<DB, U> {
-    /// Add a stage to the pipeline.
-    pub fn push<S>(mut self, stage: S) -> Self
-    where
-        S: Stage<DB> + 'static,
-    {
-        self.stages.push(QueuedStage { stage: Box::new(stage) });
-        self
-    }
-
-    /// Set the target block.
-    ///
-    /// Once this block is reached, syncing will stop.
-    pub fn with_max_block(mut self, block: Option<BlockNumber>) -> Self {
-        self.max_block = block;
-        self
-    }
-
-    /// Set a channel the pipeline will transmit events over (see [PipelineEvent]).
-    pub fn with_channel(mut self, sender: Sender<PipelineEvent>) -> Self {
-        self.events_sender.set(Some(sender));
-        self
+    /// Construct a pipeline using a [`PipelineBuilder`].
+    pub fn builder() -> PipelineBuilder<DB, U> {
+        PipelineBuilder::default()
     }
 
-    /// Set a [SyncStateUpdater].
-    pub fn with_sync_state_updater(mut self, updater: U) -> Self {
-        self.sync_state_updater = Some(updater);
-        self
+    /// Listen for events on the pipeline.
+    pub fn events(&mut self) -> mpsc::UnboundedReceiver<PipelineEvent> {
+        self.listeners.new_listener()
     }
 
     /// Run the pipeline in an infinite loop. Will terminate early if the user has specified
     /// a `max_block` in the pipeline.
     pub async fn run(&mut self, db: Arc<DB>) -> Result<(), PipelineError> {
         loop {
             let mut state = PipelineState {
-                events_sender: self.events_sender.clone(),
+                listeners: self.listeners.clone(),
```

Comment:
this looks a bit weird.

but nothing we can do in this PR about it.

looking at `self.run_loop(&mut state,` seems unnecessary to clone the listeners into the state, at least here, perhaps there are issues when it comes to the stages.